Integrating SearchBase, callbacks and renders

ShahAnuj2610 opened the pull request:
DataSearch as in reactivesearch

ShahAnuj2610 committed:
feat: add exhaustive proptypes

ShahAnuj2610 committed:
feat: add skeletion for DataSearch component

ShahAnuj2610 committed:
feat: add Error and Loader components

ShahAnuj2610 committed:
feat: add downshift with necessary components

ShahAnuj2610 committed:
feat: integrate SearchBase

jyash97 commented on src/components/DataSearch.jsx:
```diff
@@ -0,0 +1,365 @@
+      credentials
+    });
+
+    this.searchBase.subscribeToStateChanges(() => {
```
Why we are doing `forceUpdate` ?

Can we not do it with setting the state variables instead of relying on context ( `this` ) variable?

ShahAnuj2610 replied:
true

ShahAnuj2610 committed:
feat: use babel-macro

ShahAnuj2610 committed:
fix: styles

ShahAnuj2610 committed:
fix: use state instead of forceUpdate

ShahAnuj2610 committed:
feat: add source-map-explorer script

ShahAnuj2610 commented on src/components/DataSearch.jsx:
```diff
@@ -0,0 +1,380 @@
+
+    this.searchBase.subscribeToStateChanges(
+      ({ suggestions }) => {
+        this.setState({
```
@bietkul when I compare `suggestions.prev.data` and `suggestions.next.data`, they are same, seems to be a bug in `SearchBase`

jyash97 commented on src/components/DataSearch.jsx:
```diff
@@ -0,0 +1,380 @@
+      credentials
+    });
+
+    this.searchBase.subscribeToStateChanges(
```
How about using `onSuggestion` instead ?

ShahAnuj2610 replied:
`onSuggestion(prev, next)` => here also `prev` and `next` are getting the same value

jyash97 replied:
Oh, not about the data. I am asking what if we use `onSuggestion` instead of `subscribe` as the library would only focus on Suggestions.

ShahAnuj2610 replied:
> Oh, not about the data. I am asking what if we use `onSuggestion` instead of `subscribe` as the library would only focus on Suggestions.

`state management` have been done in the subscribe method and callbacks have been handled in this way, just to keep both of them independent
```js
this.searchBase.onSuggestion = props.onSuggestion
this.searchBase.onError = props.onError
```
If I had handled state in onSuggestion, then would have to handle two conditions in one function,
```js
this.searchBase.onSuggestion = (prev, next) => {
    setState({});
    if (props.onSuggestions) props.onSuggestion(prev, next)
}
```
Any specific rationale behind handling state in onSuggestion?

jyash97 replied:
Subscribe would be called on every state changes like result, pending state and other state changes but onSuggestion is only called when we get suggestions. So we render only when we get suggestion.

ShahAnuj2610 replied:
no, we can define on which properties to subscribe
![image](https://user-images.githubusercontent.com/10362927/65676960-6494ab80-e06e-11e9-94ae-ccdc5e4f5b4c.png)

jyash97 replied:
Cool. Do unsubscribe it on unmount

ShahAnuj2610 committed:
feat: handle error, results, value callbacks

ShahAnuj2610 committed:
feat: integrate SearchIcon and its event

ShahAnuj2610 committed:
feat: handle clear and suggestions callback

ShahAnuj2610 committed:
feat: handle focus callback

ShahAnuj2610 committed:
feat: add componentDidUpdate

ShahAnuj2610 committed:
fix: error and loading

ShahAnuj2610 commented:
@jyash97 this branch can be reviewed. Things that are left:
- voiceSearch
- storybook
- publish script for npm

ShahAnuj2610 committed:
fix: css emotion prop

jyash97 commented on src/addons/SuggestionItem.jsx:
```diff
@@ -0,0 +1,71 @@
+    ) : (
+      label
+    );
+  } else if (title || image || description) {
```
What will be the usage for this ?

To what I can see it is used when we need to pass image title and description from the parent component but the item that you are passing down will always be like:

```json
{
   "label" : "",
   "value": "",
   "source":{}
}
```

Do correct me if I am wrong.

jyash97 commented on src/addons/SuggestionsWrapper.jsx:
```diff
@@ -0,0 +1,22 @@
+      innerClassName || ''
+    )}`}
+  >
+    <li>{children}</li>
```
Let us not put `li` here. You can simply render inside the `div`. Currently, it's breaking the styles, you can reproduce using: 

```
renderNoSuggestion={() => <p>No Suggestion</p>}
```

jyash97 commented on src/components/Input.js:
```diff
@@ -0,0 +1,162 @@
+import { css } from '@emotion/core';
+import styled from '@emotion/styled/macro';
+
```
This file can be renamed as `styles.js` as it contains all the styles. And you can do name export for every component style

jyash97 commented on src/components/Input.js:
```diff
@@ -0,0 +1,162 @@
+      width: 100%;
+      max-height: 2.3rem;
+      line-height: 1.2rem;
+      -webkit-line-clamp: 2;
```
is this supported in Firefox and other major browser?

jyash97 commented on src/components/Input.js:
```diff
@@ -0,0 +1,162 @@
+const suggestions = (themePreset, theme) => css`
+  display: block;
+  width: 100%;
+  border: 1px solid #ccc;
```
Lets use theme here as well for proper customization, and use themes wherever we can for a better customization. You can create a JSON file with different theme config and check it.

jyash97 reviewed (changes requested):
Here are more points to take care of: 

- Lets put all styles in a single folder and if you want you can have separate files.
- Let us decrease DataSearch input height and put the position for Dropdown such that it touches the border of Input.
- Put initialization of SearchBase inside try-catch as whenever it throws error it will break the whole app. ( Reproduce using `dataField={null}`)
- Put theme wherever possible instead of hardcoding values ( even on icons )

siddharthlatest reviewed (changes requested):
Things I found based on a cursory test:
1. The component width doesn't scale down based on the viewport width,
2. The default fonts don't seem to be loading correctly,
3. Suggestion dropdown overlays on top of the input box making the text's vertical alignment look weird. See below:
 ![](https://i.imgur.com/nyuRCXZ.png)

### Recommendations:
1. Rename the component to `SearchBox`,
2. Rename `index` prop to `app`. `url` prop should be optional (defaults to `https://scalr.api.appbase.io`),
3. We should have a `URLParams` prop. Are there other props present in DataSearch that aren't present here?,
4. Other prop inconsistenencies with naming convention should be corrected, e.g. `enableVoiceSearch` here is `showVoiceSearch` in DataSearch / ReactiveSearch. We should be consistent.

ShahAnuj2610 commented:
@siddharthlatest 
> 1. The component width doesn't scale down based on the viewport width,

Seems to be some issue with the build, working perfectly in my local
![image](https://user-images.githubusercontent.com/10362927/65854854-4b943f00-e37b-11e9-8954-09d95c71c0f1.png)


> 3. We should have a `URLParams` prop. Are there other props present in DataSearch that aren't present here?,

Can you share me the use-case, how will this be used?

> 4. Other prop inconsistenencies with naming convention should be corrected, e.g. `enableVoiceSearch` here is `showVoiceSearch` in DataSearch / ReactiveSearch. We should be consistent.

we are not only showing voice search, but we are also enabling its functionality, so thought this prop name seems more meaningful

ShahAnuj2610 committed:
refactor: files and remove unecessary component

ShahAnuj2610 committed:
fix: webkit--line-clamp

ShahAnuj2610 committed:
fix: styles

ShahAnuj2610 committed:
fix: handle SearchBase initialization inside try catch

jyash97 commented:
Does this not include rollup config ?

jyash97 reviewed (commented):
Here are my points after checking the App: 

- On Mobile Check this some style issue: 
 
![image](https://user-images.githubusercontent.com/22376783/65890054-eb73bc00-e3bf-11e9-8dc4-6cb238bafe81.png)

- After adding `loader` prop. I see it working only in initial fetch after that it doesn't work.

- In `render` prop the `loading` is always `undefined`: 
![image](https://user-images.githubusercontent.com/22376783/65890471-8f5d6780-e3c0-11e9-89ba-88cc0da39049.png)

- If we use render prop we won't need to show the normal loader

- `No suggestion` and loader messages are aligned to the center. Lets remove this style for now.

- Adding `theme` prop breaks the UI: 
![image](https://user-images.githubusercontent.com/22376783/65891112-a2bd0280-e3c1-11e9-8623-94bbb3968b7a.png)

`Theme Value`: 
```
theme={{
     colors: {
            textColor: 'red',
            primaryTextColor: 'blue',
            primaryColor: 'green',
            titleColor: 'yellow',
            alertColor: 'aquamarine'
          }
 }}
```

- Putting `themePreset="dark"` breaks the UI

- Rename Input.js to styles.js

- Adding `title` prop breaks the UI. The reason is that you are not passing `theme` prop.

- The title should not be center aligned

Sandbox Link that I was checking: https://codesandbox.io/s/react-searchbox-wd9y4

ShahAnuj2610 commented:
> Here are my points after checking the App:
> 
> * On Mobile Check this some style issue:
> 
> ![image](https://user-images.githubusercontent.com/22376783/65890054-eb73bc00-e3bf-11e9-8dc4-6cb238bafe81.png)

fixed [here](https://github.com/appbaseio/react-searchbox/pull/2)

> * After adding `loader` prop. I see it working only in initial fetch after that it doesn't work.

seems to be an issue with `SearchBase`, will have to debug there

> * In `render` prop the `loading` is always `undefined`:
>   ![image](https://user-images.githubusercontent.com/22376783/65890471-8f5d6780-e3c0-11e9-89ba-88cc0da39049.png)

related to the above issue

> * If we use render prop we won't need to show the normal loader

fixed [here](https://github.com/appbaseio/react-searchbox/pull/2)

> * `No suggestion` and loader messages are aligned to the center. Lets remove this style for now.

fixed [here](https://github.com/appbaseio/react-searchbox/pull/2)

> * Adding `theme` prop breaks the UI:
>   ![image](https://user-images.githubusercontent.com/22376783/65891112-a2bd0280-e3c1-11e9-8623-94bbb3968b7a.png)
> 
> `Theme Value`:
> 
> ```
> theme={{
>      colors: {
>             textColor: 'red',
>             primaryTextColor: 'blue',
>             primaryColor: 'green',
>             titleColor: 'yellow',
>             alertColor: 'aquamarine'
>           }
>  }}
> ```
> 
> * Putting `themePreset="dark"` breaks the UI

fixed [here](https://github.com/appbaseio/react-searchbox/pull/2)

> * Rename Input.js to styles.js

fixed [here](https://github.com/appbaseio/react-searchbox/pull/2)

> * Adding `title` prop breaks the UI. The reason is that you are not passing `theme` prop.
> * The title should not be center aligned
> 

fixed [here](https://github.com/appbaseio/react-searchbox/pull/2)

> Sandbox Link that I was checking: https://codesandbox.io/s/react-searchbox-wd9y4

@jyash97 This PR can be merged, changes are done in the subsequent PR
